### PR TITLE
FEATURE: Nested Tags

### PR DESCRIPTION
**Pull request by daniellienert**
This adds a naive implementation of nested tags using a parent field. 
It is intentionally not exposed in the UI, as nested Tags will be a feature of the new Media UI only.

#solves: https://github.com/Flowpack/media-ui/issues/3

### CODE REVIEW

**Comment by daniellienert** [comment]
Getting the nested structure with a single recursive query could be done like that:
```
with recursive taghierarchy (persistence_object_identifier, label, parent) as (
              select     persistence_object_identifier, label, parent
              from       neos_media_domain_model_tag
              where      persistence_object_identifier = ?
              union all
              select     t.persistence_object_identifier, t.label, t.parent
              from       neos_media_domain_model_tag t
              inner join taghierarchy
                      on t.parent = taghierarchy.persistence_object_identifier
            )
            select * from taghierarchy;
```

Question is, should that be part of the Tags Repository or should we put that to a class in the Media UI?

**Comment by Sebobo** [comment]
The answer is related to whether we want a nested structure in our response. With GraphQL you don‘t have recursive data structures (or you shouldn‘t have) as far as I understood. 
Therefore getting a flat list of tags and reconstructing the tree would be the basic approach.
So I think its fine to put it into the repository but the method will not be a necessity for the Media UI.

**Comment by daniellienert** [comment]
Ok, if the levels are fetched iteratively it's not needed.

**Comment by Sebobo** [comment]
Do we want something like a max nesting depth?

**Comment by daniellienert** [comment]
@Sebobo I see no reason why we should limit that from a backend code persepective. Depends on the frontend code how many levels can be handled, so I would limit it there if needed.

File: Neos.Media/Classes/Domain/Model/Tag.php [hunk 1]
```diff
@@ -99,4 +116,40 @@ public function setAssetCollections(Collection $assetCollections)
         }
         $this->assetCollections = $assetCollections;
     }
+
+    /**
+     * @return Tag
```
Comment (albe) [inline Neos.Media/Classes/Domain/Model/Tag.php]:
```suggestion
     * @return Tag|null
```

**Comment by albe** [comment]
So is there anything left to do right now?

File: Neos.Media/Classes/Domain/Model/Tag.php [hunk 2]
```diff
@@ -1,4 +1,5 @@
 <?php
+declare(strict_types=1);
```
[the remark this answers is not in the document]
Comment (daniellienert) [reply Neos.Media/Classes/Domain/Model/Tag.php:2]: I'm ok with removing it.

**Comment by albe** [comment]
So travis is just messing up on that one behat test. Restarted it, but it still fails with the git checkout error. Will merge anyhow.

**Pull request merged**